File: 200-number-of-islands/cpp/main.cpp
class Solution {
 public:
  int numIslands(vector<vector<char>>& grid) {
    int count = 0;
    for (int i = 0; i < grid.size(); ++i) {
      for (int j = 0; j < grid[i].size(); ++j) {
        count += bfs(grid, i, j);
      }
    }

    return count;
  }

  int bfs(vector<vector<char>>& grid, int i, int j) {
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[i].size() ||
        grid[i][j] == '0') {
      return 0;
    }

    grid[i][j] = '0';
    bfs(grid, i - 1, j);
    bfs(grid, i + 1, j);
    bfs(grid, i, j - 1);
    bfs(grid, i, j + 1);
    return 1;
  }
};

File: 200-number-of-islands/go/main.go
package main

func numIslands(grid [][]byte) int {
	count := 0
	for i, _ := range grid {
		for j, _ := range grid[i] {
			count += bfs(&grid, i, j)
		}
	}
	return count
}

func bfs(grid *[][]byte, i int, j int) int {
	if i < 0 || j < 0 || i >= len(*grid) || j >= len((*grid)[i]) || (*grid)[i][j] == uint8('0') {
		return 0
	}

	(*grid)[i][j] = uint8('0')

	bfs(grid, i-1, j)
	bfs(grid, i+1, j)
	bfs(grid, i, j-1)
	bfs(grid, i, j+1)

	return 1
}
